dG (2D) - Mesh Adjacency Map

int-ptr-ptr opened the pull request:
## Description

Implements an adjacency map in the `mesh` namespace. The adjacency map takes a `mesh::mesh` and stores the conforming and nonconforming adjacencies of that mesh.

On completion, one should be able to

1. Load a mesh (potentially nonconforming) database generated in `meshfem2d`
2. Generate (and manipulate) an adjacency map storing `(ispec, edge) -> (ispec,edge) or mortar anchor or boundary`
3. Build a `compute::mesh` from the adjacency map (without mortars for now), respecting the adjacencies from the map.

Once nonconforming meshes are implemented, the mortars and flux scheme parameters should be loaded during / prior to (3), changing the adjacency map accordingly.

## Issue Number

#846


## Checklist

Please make sure to check developer documentation on specfem docs.

- [x] I ran the code through pre-commit to check style
- [ ] My code passes all the integration tests
- [ ] I have added sufficient unittests to test my changes
- [ ] I have added/updated documentation for the changes I am proposing
- [x] I have updated CMakeLists to ensure my code builds
- [ ] My code builds across all platforms

## Features

- [x] Mapping `(ispec, edge)` to a conforming adjacent `(ispec, edge)`.
- [x] Sampling if `(ispec, edge)` is on the boundary. (Currently, there is no way to tell if a boundary is Neumann, so it and the mortars will be mixed-up.)
- [x] Adjacency map auto-connects conforming elements that share control node indices on construction. For now, the adjacency map does not consider two control nodes at the same position as the same. I think it would be best to keep it this way, but a future PR can change it if desired.
- [x] Creation of an index mapping `(ispec,iz,ix) -> iglob` that respects the adjacency map rather than the node positions.

### Tests

- [x] Adjacency map generation
  - [x] small conforming mesh
- [x] Adjacency map to index mapping
  - [x] index mapping validity (surjective, etc)

int-ptr-ptr committed:
adjmap first pass, started test

int-ptr-ptr committed:
adjacency map -> index mapping

int-ptr-ptr committed:
Merge branch 'devel' into adjacency2d

Rohit-Kakodkar reviewed (commented):
Adjacency map for conforming elements is already built by the mesher. I think we should write that to the database and use it. https://github.com/PrincetonUniversity/SPECFEMPP/blob/638767a2609aa3c460ff695446a850e1ee49d07a/fortran/meshfem2d/decompose_mesh.F90#L66 

This is should also account for partitioned mesh when using MPI. Since the databases would be separate for each individual processor.

Rohit-Kakodkar commented:
To address MPI concerns effectively, we should focus on creating non-conforming adjacency within meshfem2d. Since the mesher has access to all partitions, we can avoid the complications associated with reductions.

int-ptr-ptr committed:
use meshfem adj, adjmap part of mesh struct

+ added optional footers to meshfem database
+ adjacency map can now build from meshfem database
+ migrated adjacency_map into mesh::mesh. Can be uninitialized.
+ added both footer and no-footer cases to adjtest_2d.

int-ptr-ptr committed:
Merge remote-tracking branch 'origin/devel' into adjacency2d

int-ptr-ptr committed:
test -> fixture, de-macro adjacency code

int-ptr-ptr committed:
assign_numbering delegates to adjmap if it can

int-ptr-ptr commented on src/io/mesh/impl/fortran/dim2/mesh.cpp:
```diff
@@ -136,6 +137,12 @@ specfem::mesh::mesh<specfem::dimension::type::dim2> specfem::io::read_2d_mesh(
   }
 
+  try {
+    specfem::io::mesh::impl::fortran::dim2::read_footer(stream, mesh, mpi);
```
In order to minimize the burden on users and test / example codes, I have the database export an optional "footer". If after `axial_elements`, an `EOF` is read, then SPECFEM knows there is no footer. Otherwise, for the sake of backwards compatibility (based on when databases were generated), a version code (3 `int`s) is output, followed by labeled footer entries. For now, there is just:
- [`0`](https://github.com/PrincetonUniversity/SPECFEMPP/blob/42ade20928e642e1012f859c95200f0811049b4b/fortran/meshfem2d/save_databases_footer.f90#L11) - end the footer. If we do not see `EOF` at [this point](https://github.com/PrincetonUniversity/SPECFEMPP/blob/42ade20928e642e1012f859c95200f0811049b4b/src/io/mesh/impl/fortran/dim2/mesh.cpp#L146-L150) after the footer-read, something went wrong.
- [`1`](https://github.com/PrincetonUniversity/SPECFEMPP/blob/42ade20928e642e1012f859c95200f0811049b4b/fortran/meshfem2d/save_databases_adj.f90#L12) - the adjacency map is written next. The format is an inter-woven `xadj` and `adjncy`:
```
//element 0 adjacencies:
num_adj0 // ( = (xadj_g(1) - xadj_g(0)) )
adjncy(0)
...
adjncy(num_adj0-1)
//element 1 adjacencies:
num_adj1 // ( = (xadj_g(2) - xadj_g(1)) )
adjncy(num_adj0)
...
adjncy(num_adj0+num_adj1-1)
// so on to nspec
...
```
These are ints specified [here](https://github.com/PrincetonUniversity/SPECFEMPP/blob/42ade20928e642e1012f859c95200f0811049b4b/src/io/mesh/impl/fortran/dim2/read_footer.cpp#L50-L55) in specfem.

Rohit-Kakodkar replied:
Thanks, I see now whats going on. I'd suggest a few changes to the workflow so things are consistent with `meshfem2d`. Since meshfem databases dont follow a version for backward compatibility, I am reluctant to include it now. 

- Create a parameter within par file called `WRITE_ADJECENCY_MAP = true/false`
- Write the above parameter within the database. This will break backward compatibility in the reader. We can fix this using `try catch` block for now until we update the current databases using new mesher. 
- Write the adjacency database only if the above parameter is set to true
- Read the database if the parameter is true

```cpp

  bool read_adjacency_mapping;
  try {
    specfem::io::fortran_read_line(
        stream, &read_adjacency_mapping);
  catch (std::runtime_error &e) {
    read_adjacency_mapping = false;
  }

  if (read_adjacency_mapping) {
    try {
      mesh.adjacency_mapping =
          specfem::io::mesh::impl::fortran::dim2::read_adjacency_mapping(
              stream, mesh.parameters.nspec, mesh.control_nodes.knods, mpi);
    } catch (std::runtime_error &e) {
      throw;
    }
  }

```

This should give us sometime to update the test databases after this gets merged. 

One thing to note here is - since you're updating how adjacency map is computed. I'd add a test within `displacement-test` that runs a simulation for a database with adjacency map. You could copy one of the tests and then change Par file so that code path uses the adjacency map.

int-ptr-ptr replied:
- [adjacency map write parameter](https://github.com/PrincetonUniversity/SPECFEMPP/blob/3fa296accf6625db5a4e1d6ddeff10d936055dec/fortran/meshfem2d/read_parameter_file.F90#L540)
- [displacement-tests test](https://github.com/PrincetonUniversity/SPECFEMPP/blob/3fa296accf6625db5a4e1d6ddeff10d936055dec/tests/unit-tests/displacement_tests/Newmark/test_config.yaml#L111-L118)

lsawade commented on src/mesh/dim2/adjacency_map/relative_nodes_inds_generator.py:
```diff
@@ -0,0 +1,63 @@
+import pathlib
```
NO! We do not want a true python dependency within the C++ code! 

I think this should be done in a way that does not require python code generation.

icui replied:
I think it's better to keep the Python code as comments in `relative_node_inds.cpp`.

lsawade commented on tests/unit-tests/mortar/runner.cpp:
```diff
@@ -0,0 +1,10 @@
+#include "../MPI_environment.hpp"
+#include "gtest/gtest.h"
+
+int main(int argc, char *argv[]) {
+  ::testing::InitGoogleTest(&argc, argv);
+  ::testing::AddGlobalTestEnvironment(new MPIEnvironment);
+  ::testing::AddGlobalTestEnvironment(new KokkosEnvironment);
+  return RUN_ALL_TESTS();
+}
```
Was not able to attached a comment to `database.bin` but if you create a test that requires a generated database, you need to also provide the provenance, that is how did you create `database.bin`

int-ptr-ptr replied:
- [provenance added](https://github.com/PrincetonUniversity/SPECFEMPP/pull/851/files#diff-7773cd3d46219df71a1135a94050fe78d4c7778e08f5268f9ef37175bcbb4f5a)

lsawade reviewed (changes requested):
This already looks pretty great! I'm requesting a couple of changes:
- Remove versioning of functions
- Remove code generation through python. A quick prompt to chatgpt the like will give you a plethora of options that don't require python to generate code.
- add provenance to the generation of `database.bin`

int-ptr-ptr committed:
removed footer versioning, added provenance

int-ptr-ptr committed:
added meshfem parameter write_adjacency_map

int-ptr-ptr committed:
adjmap ngnod = 4 case

int-ptr-ptr committed:
added displacement-test + bugfix

int-ptr-ptr commented on src/mesh/dim2/adjacency_map/adjacency_map.cpp:
```diff
@@ -0,0 +1,381 @@
+    specfem::dimension::type::dim2>::adjacency_map()
+    : nspec(-1) {}
+
+static inline int node_edge(const specfem::enums::edge::type &edgetype,
```
[Now accounts for `ngnod = 4`](https://github.com/PrincetonUniversity/SPECFEMPP/blob/3fa296accf6625db5a4e1d6ddeff10d936055dec/src/mesh/dim2/adjacency_map/adjacency_map.cpp#L155-L161)

int-ptr-ptr commented on fortran/meshfem2d/save_databases_footer.f90:
```diff
@@ -0,0 +1,23 @@
+  integer, parameter :: completion_footercode = 0
+
+  write(IOUT) version(:)
+  write(IMAIN,"(A,I0,A,I0,A,I0)") "  database footer: v", version(1) , ".", version(2), ".", version(3)
```
- versioning has been removed. I still have the `footercode`s in place for `specfem` to know if the adjacency map was stored, however. It makes more sense to have `load_adjacency_map` as a database-stored parameter than a `specfem_config.yaml` parameter, as per Rohit's suggestion.

> • Create a parameter in the meshfem2d parameter file: Add WRITE_ADJACENCY_MAP = true/false.
> • Store the parameter in the database: This change will affect backward compatibility in the reader. We can use a try-catch block for now until we update the databases with the new mesher.
> • Write the adjacency database only if the parameter is true.

Rohit-Kakodkar commented:
Hi Kentaro, Thanks for making the changes. Can you add doc strings for functions where you can. We are very busy this week with a new release. We are planning to review this next tuesday.

int-ptr-ptr committed:
added docstrings

int-ptr-ptr committed:
Merge branch 'devel' into adjacency2d

int-ptr-ptr committed:
updated adjmap test to current suite

int-ptr-ptr committed:
fixed adjmap test

lsawade commented on fortran/meshfem2d/save_databases_footer.f90:
```diff
@@ -0,0 +1,22 @@
+
+  implicit none
+
+  integer, parameter :: completion_footercode = 0
```
If you want to separate out the additional mesh things, I suggest to do something like `save_database_nonconform.f90` and then put all the relevant functions into a single module with one main function that calles `write_non_conform_parameters()`.

Additionally, I think it would make sense to write a boolean that says you are writing the `non_conform_database`.

int-ptr-ptr replied:
We should talk about this tomorrow.

lsawade commented on fortran/meshfem2d/shared_par.f90:
```diff
@@ -299,6 +299,12 @@ module shared_input_parameters
+   !# Database output parameters
+   !#
+   !#-----------------------------------------------------------------------------
+   logical :: write_adjacency_map
```
This is the parameter i mean. Write it to the database

Rohit-Kakodkar commented on src/mesh/dim2/adjacency_map/adjacency_map.cpp:
```diff
@@ -0,0 +1,401 @@
+
+        // iterate over candidates found by meshfem
+        for (int ispec_o : elements_with_shared_nodes[ispec]) {
+          if (ispec_o > ispec) {
+            // we we only need to iterate adjacencies once.
+            continue;
+          }
+
+          for (auto edge_o : {
+                   specfem::enums::edge::type::RIGHT,
+                   specfem::enums::edge::type::TOP,
+                   specfem::enums::edge::type::LEFT,
+                   specfem::enums::edge::type::BOTTOM,
+               }) {
+            matching = true;
+            // adjacent element has same nodes, but in clockwise direction
+            for (int i = 0; i < 3; i++) {
+              if (i == 1 && parent.control_nodes.ngnod == 4) {
+                // no middle node for ngnod = 4
+                continue;
+              }
+              if (nodes_edge[i] != parent.control_nodes.knods(
+                                       node_edge(edge_o, 2 - i), ispec_o)) {
+                matching = false;
+                break;
+              }
+            }
+            if (matching) {
+              create_conforming_adjacency(ispec, edge, ispec_o, edge_o);
+              break;
+            }
+          }
+          if (matching) {
+            break;
+          }
+        }
+      }
+    }
+  }
+}
```
This loop should be split into 2 cases of 4 control nodes and 9 control nodes. It would make it much easier to read.

Rohit-Kakodkar commented on src/mesh/dim2/adjacency_map/adjacency_map.cpp:
```diff
@@ -0,0 +1,401 @@
+            continue;
+          }
+
+          for (auto edge_o : {
```
Can we make this into a function `find_conforming_edge`?

lsawade commented on src/mesh/dim2/adjacency_map/adjacency_map.cpp:
```diff
@@ -0,0 +1,401 @@
+
+specfem::mesh::adjacency_map::adjacency_map<specfem::dimension::type::dim2>::
+    adjacency_map(
+        const specfem::mesh::mesh<specfem::dimension::type::dim2> &parent,
```
why is this called parent?

Rohit-Kakodkar commented on src/mesh/dim2/adjacency_map/adjacency_map.cpp:
```diff
@@ -0,0 +1,401 @@
+          specfem::enums::edge::type::LEFT, true>(*this, adj, search, bd);
+      break;
+    case enums::boundaries::type::BOTTOM_LEFT:
+      get_all_conforming_adjacencies__test_corner<
+          specfem::enums::edge::type::LEFT, false>(*this, adj, search, bd);
+      get_all_conforming_adjacencies__test_corner<
+          specfem::enums::edge::type::BOTTOM, true>(*this, adj, search, bd);
+      break;
+    case enums::boundaries::type::BOTTOM_RIGHT:
+      get_all_conforming_adjacencies__test_corner<
+          specfem::enums::edge::type::BOTTOM, false>(*this, adj, search, bd);
+      get_all_conforming_adjacencies__test_corner<
+          specfem::enums::edge::type::RIGHT, true>(*this, adj, search, bd);
+      break;
+    case enums::boundaries::type::TOP_RIGHT:
+      get_all_conforming_adjacencies__test_corner<
+          specfem::enums::edge::type::RIGHT, false>(*this, adj, search, bd);
+      get_all_conforming_adjacencies__test_corner<
+          specfem::enums::edge::type::TOP, true>(*this, adj, search, bd);
+      break;
+    case enums::boundaries::type::TOP:
+      get_all_conforming_adjacencies__test_edge<
+          specfem::enums::edge::type::TOP>(*this, adj, search, bd);
+      break;
+    case enums::boundaries::type::LEFT:
+      get_all_conforming_adjacencies__test_edge<
+          specfem::enums::edge::type::LEFT>(*this, adj, search, bd);
+      break;
+    case enums::boundaries::type::RIGHT:
+      get_all_conforming_adjacencies__test_edge<
+          specfem::enums::edge::type::RIGHT>(*this, adj, search, bd);
+      break;
+    case enums::boundaries::type::BOTTOM:
+      get_all_conforming_adjacencies__test_edge<
+          specfem::enums::edge::type::BOTTOM>(*this, adj, search, bd);
+      break;
+    }
+  }
+  return adj;
+}
```
What are these being used for?

Rohit-Kakodkar commented on src/mesh/dim2/adjacency_map/adjacency_map.cpp:
```diff
@@ -0,0 +1,401 @@
+  switch (bd) {
+  case specfem::enums::edge::TOP:
+    return specfem::enums::boundaries::type::TOP;
+  case specfem::enums::edge::BOTTOM:
+    return specfem::enums::boundaries::type::BOTTOM;
+  case specfem::enums::edge::LEFT:
+    return specfem::enums::boundaries::type::LEFT;
+  case specfem::enums::edge::RIGHT:
+    return specfem::enums::boundaries::type::RIGHT;
+    break;
+  default:
+    // this should never be called
+    return specfem::enums::boundaries::type::RIGHT;
+  }
+}
+// ====================================================================
+// helper: given an edge, returns either the clockwise or counterclockwise
+// corner.
+static inline specfem::enums::boundaries::type
+edge_and_polarity_to_corner(const specfem::enums::edge::type &bd,
+                            const bool counterclockwise) {
+  switch (bd) {
+  case specfem::enums::edge::TOP:
+    return counterclockwise ? specfem::enums::boundaries::type::TOP_LEFT
+                            : specfem::enums::boundaries::type::TOP_RIGHT;
+  case specfem::enums::edge::BOTTOM:
+    return counterclockwise ? specfem::enums::boundaries::type::BOTTOM_RIGHT
+                            : specfem::enums::boundaries::type::BOTTOM_LEFT;
+  case specfem::enums::edge::LEFT:
+    return counterclockwise ? specfem::enums::boundaries::type::BOTTOM_LEFT
+                            : specfem::enums::boundaries::type::TOP_LEFT;
+  case specfem::enums::edge::RIGHT:
+    return counterclockwise ? specfem::enums::boundaries::type::TOP_RIGHT
+                            : specfem::enums::boundaries::type::BOTTOM_RIGHT;
+    break;
+  default:
+    // this should never be called
+    return specfem::enums::boundaries::type::RIGHT;
+  }
+}
```
Are these being used?

int-ptr-ptr replied:
`bdtype_to_edge` is used when [reading in boundaries and marking the corresponding edge in `adjacency_map`](https://github.com/int-ptr-ptr/SPECFEMPP/blob/adjacency2d/src/mesh/dim2/adjacency_map/adjacency_map.cpp#L78-L85).
```c++
static inline void boundarymark(
    specfem::mesh::adjacency_map::adjacency_map<specfem::dimension::type::dim2>
        &adjmap,
    const specfem::mesh::mesh<specfem::dimension::type::dim2> &parent) {
  for (int i = 0; i < parent.boundaries.absorbing_boundary.nelements; i++) {
    const auto edgetype =
        bdtype_to_edge(parent.boundaries.absorbing_boundary.type(i));
    if (edgetype != specfem::enums::edge::type::NONE) {
      adjmap.set_as_boundary(
          parent.boundaries.absorbing_boundary.index_mapping(i), edgetype);
    }
  }
  for (int i = 0;
       i < parent.boundaries.acoustic_free_surface.nelem_acoustic_surface;
       i++) {
    const auto edgetype =
        bdtype_to_edge(parent.boundaries.acoustic_free_surface.type(i));
    if (edgetype != specfem::enums::edge::type::NONE) {
      adjmap.set_as_boundary(
          parent.boundaries.acoustic_free_surface.index_mapping(i), edgetype);
    }
  }
  // forcing boundary is never used, so skip it.
}
```

`all_conforming_adjacencies()` is a set of boundaries, so for edges, we convert them with `edge_to_bdtype`:
```c++
template <specfem::enums::edge::type edgetype>
static inline void get_all_conforming_adjacencies__test_edge(
    const specfem::mesh::adjacency_map::adjacency_map<
        specfem::dimension::type::dim2> &map,
    std::set<std::pair<int, specfem::enums::boundaries::type> > &adj,
    std::list<std::pair<int, specfem::enums::boundaries::type> > &search,
    std::pair<int, specfem::enums::boundaries::type> &front) {
  /* helper for get_all_conforming_adjacencies:
   * given a corner (specified by edge and direction), try_append the opposite
   * corner (across the given mating edge).
   */
  if (map.has_conforming_adjacency(
          front.first,
          specfem::mesh::adjacency_map::adjacency_map<
              specfem::dimension::type::dim2>::edge_to_index(edgetype))) {
    const auto other = map.get_conforming_adjacency(
        front.first,
        specfem::mesh::adjacency_map::adjacency_map<
            specfem::dimension::type::dim2>::edge_to_index(edgetype));
    get_all_conforming_adjacencies__try_append(adj, search, other.first,
                                               edge_to_bdtype(other.second));
  }
}
```

The corner case requires the polarity (ccw, cw) to find the shared side:
```c++
template <specfem::enums::edge::type edgetype, bool is_ccw_from_corner>
static inline void get_all_conforming_adjacencies__test_corner(
    const specfem::mesh::adjacency_map::adjacency_map<
        specfem::dimension::type::dim2> &map,
    std::set<std::pair<int, specfem::enums::boundaries::type> > &adj,
    std::list<std::pair<int, specfem::enums::boundaries::type> > &search,
    std::pair<int, specfem::enums::boundaries::type> &front) {
  /* helper for get_all_conforming_adjacencies:
   * given a corner (specified by edge and direction), try_append the opposite
   * corner (across the given mating edge).
   */
  if (map.has_conforming_adjacency(
          front.first,
          specfem::mesh::adjacency_map::adjacency_map<
              specfem::dimension::type::dim2>::edge_to_index(edgetype))) {
    const auto other = map.get_conforming_adjacency(
        front.first,
        specfem::mesh::adjacency_map::adjacency_map<
            specfem::dimension::type::dim2>::edge_to_index(edgetype));
    get_all_conforming_adjacencies__try_append(
        adj, search, other.first,
        // edge is CCW from corner -> corner is not CCW side of edge
        // double negative from the polarity flip
        edge_and_polarity_to_corner(other.second, is_ccw_from_corner));
  }
}
```

Rohit-Kakodkar reviewed (commented):
This PR is way too long for us to be able to review it altogether. I am trying to review it in piece by piece. Please address the comments below.

lsawade commented on src/mesh/dim2/adjacency_map/adjacency_map.cpp:
```diff
@@ -0,0 +1,401 @@
+  {
+    for (int ispec = 0; ispec < nspec; ispec++) {
+
+      int nodes_edge[3]; // store nodes along edge, in counterclockwise
```
Ar there not just 2 nodes on the edge when ngnod==4?

lsawade commented on src/mesh/dim2/adjacency_map/adjacency_map.cpp:
```diff
@@ -0,0 +1,401 @@
+           }) {
+        // did we find the candidate to this edge?
+        bool matching = false;
+        // populate nodes_edge according to which edge this is
+        for (int i = 0; i < 3; i++) {
+          if (i == 1 && parent.control_nodes.ngnod == 4) {
+            // no middle node for ngnod = 4
+            continue;
+          }
+          nodes_edge[i] = parent.control_nodes.knods(node_edge(edge, i), ispec);
+        }
```
Asking about the above, since this for loop is odd. You are not populating nodes_edge[1] for ngnod 4, Which seems weird

lsawade reviewed (commented):
Minor comments, looks mostly good to me. 

One major one, I would probably write the `write_adjacency` try reading that and only read the adjacency if that `write_adjacency` is there and `true`. Then you don't have to do the whole footer business and check which footer type etc.

int-ptr-ptr committed:
changes for PR approval

icui commented on fortran/meshfem2d/decompose_mesh.F90:
```diff
@@ -88,7 +88,7 @@ subroutine decompose_mesh()
     enddo
 
-    if (NPROC > 1) then
+    if (NPROC >= 1) then
```
This if condition is redundant.

icui commented on include/mesh/dim2/adjacency_map/adjacency_map.hpp:
```diff
@@ -0,0 +1,213 @@
+   * index from an edge::type -- this is set explicitly since it must not
+   * change, and some logic is simplified with this counter-clockwise ordering.
+   */
+  static constexpr specfem::enums::edge::type edge_from_index(const int edge) {
```
Why did you use a new numbering when the edge is already defined in specfem as 
```cpp
namespace edge {
/**
 * @brief Edge type enumeration
 *
 */
enum type {
  NONE = 0,   /// Not an edge
  TOP = 1,    ///< Top edge
  BOTTOM = 2, ///< Bottom edge
  LEFT = 3,   ///< Left edge
  RIGHT = 4   ///< Right edge
};
...
```

icui commented on include/mesh/dim2/adjacency_map/adjacency_map.hpp:
```diff
@@ -0,0 +1,213 @@
+      return 2;
+    case specfem::enums::edge::BOTTOM:
+      return 3;
+    default:
```
You should throw an error if the behavior is unexpected.

icui commented on src/compute/compute_mesh.cpp:
```diff
@@ -293,10 +322,13 @@ specfem::compute::mesh::mesh(
   this->ngllz = this->quadratures.gll.N;
 
-  this->points = this->assemble();
+  this->points = this->assemble(adjacency_map, this->mapping);
```
Why do you pass `adjacency_map` and `mapping` as parameters when they can be directly accessed by `this->assemble`

icui commented on src/mesh/dim2/adjacency_map/adjacency_map.cpp:
```diff
@@ -0,0 +1,472 @@
+// conversions between edge and boundary enums.
+
+static inline specfem::enums::edge::type
+bdtype_to_edge(const specfem::enums::boundaries::type &bd) {
```
If you could define the correct numbers in `include/enumerations/specfem_enums.hpp` for edge and boundary, functions `detype_to_edge` and `edge_to_bdtype` can be simplified.

icui commented on src/mesh/dim2/adjacency_map/adjacency_map.cpp:
```diff
@@ -0,0 +1,472 @@
+ * @brief calls set_as_boundary() for all boundary edges tagged by the
+ * mesh::mesh.
+ */
+static inline void boundarymark(
```
you should follow the snake_case naming convention unless there's any specific reasons.

lsawade reviewed (approved):
Tests look good to me. 

Main comment, if you want to print something "complex", which you can, but may not be required, you need comment what you want the result to be.

Rohit-Kakodkar commented on src/mesh/dim2/adjacency_map/adjacency_map.cpp:
```diff
@@ -126,11 +127,131 @@ static inline int node_edge(const specfem::enums::edge::type &edgetype,
+    const std::vector<int> &candidates) {
+
+  int nodes_edge[3]; // store nodes along edge, in counterclockwise
+                     // direction
+  for (int i = 0; i < 3; i++) {
+    nodes_edge[i] = mesh.control_nodes.knods(node_edge(edge, i), ispec);
+  }
+
+  // iterate over candidates found by meshfem
+  for (int ispec_o : candidates) {
+    if (ispec_o > ispec) {
+      // we we only need to iterate adjacencies once.
+      continue;
+    }
+
+    for (auto edge_o : {
+             specfem::enums::edge::type::RIGHT,
+             specfem::enums::edge::type::TOP,
+             specfem::enums::edge::type::LEFT,
+             specfem::enums::edge::type::BOTTOM,
+         }) {
+      // adjacent element has same nodes, but in clockwise direction
+      bool matching = true;
+      // adjacent element has same nodes, but in clockwise direction
+      for (int i = 0; i < 3; i++) {
+        if (i == 1 && mesh.control_nodes.ngnod == 4) {
+          // no middle node for ngnod = 4
+          continue;
+        }
+        if (nodes_edge[i] !=
+            mesh.control_nodes.knods(node_edge(edge_o, 2 - i), ispec_o)) {
+          matching = false;
+          break;
+        }
+      }
+      if (matching) {
+        return std::make_pair(ispec_o, edge_o);
+      }
+    }
+  }
```
I think this should be split up into 4 and 9 nodes.

Rohit-Kakodkar commented on include/mesh/dim2/adjacency_map/adjacency_map.hpp:
```diff
@@ -0,0 +1,213 @@
+ * @brief Stores the adjacencies between elements.
+ *
+ */
+template <specfem::dimension::type dimtype> struct adjacency_map;
```
Sorry, latest PR wasnt showing up on my end. Marking this as resolved again.

Rohit-Kakodkar commented:
okay to test

int-ptr-ptr committed:
sunset footer for adjacency map entry

int-ptr-ptr commented on tests/unit-tests/displacement_tests/Newmark/serial/HomogeneousElasticAnisotropicDomainAdjmapAssembly/provenance/Par_file:
```diff
@@ -0,0 +1,461 @@
+#-----------------------------------------------------------
```
I've been running xmeshfem2D.

int-ptr-ptr committed:
requested changes

+ always write `write_adjacency_map` flag, even if false
+ conforming adjacency provenance
+ new test for conforming_adjacencies, since assembly_mapping depends on them
+ adjacency map print function migrated to class member.
- removed mapping argument in `assemble()`
- removed the printing of the adjacency map in `run_test_conforming()`.

Rohit-Kakodkar reviewed (changes requested):
Few comments. I think you should stop adding new tests or implementations within this PR. The new tests should be added in future PR. Otherwise we'd keep on going back and forth in this PR and never get it merged.

int-ptr-ptr committed:
Merge remote-tracking branch 'origin/devel' into adjacency2d

int-ptr-ptr committed:
fixed compilation after merge

int-ptr-ptr committed:
rework (conforming) mortar test

- removed adjacencies / corners from fixture
+ adjacency test now additionally verifies against coordinates
+ added conforming-box test: a simple 10x10 grid.

int-ptr-ptr commented on src/mesh/dim2/adjacency_map/adjacency_map.cpp:
```diff
@@ -470,3 +477,92 @@ specfem::mesh::adjacency_map::adjacency_map<specfem::dimension::type::dim2>::
 }
+
+std::string
+specfem::mesh::adjacency_map::adjacency_map<specfem::dimension::type::dim2>::
```
I've commented this code out. We may have a test in the future that uses it, but the comments can be removed in the future if that's not the case.

int-ptr-ptr commented on tests/unit-tests/mortar/adjacency_map/adjtest_2d.cpp:
```diff
@@ -13,105 +13,9 @@
 #include "mesh/dim2/adjacency_map/adjacency_map.hpp"
 #include "mortar/fixture/mortar_fixtures.hpp"
 
+#define TEST_ASSEMBLY_MAPPING_DEFAULT_NGLL (5)
```
Magic number. I switched it to a `static` value to free the macro namespace.

int-ptr-ptr commented on tests/unit-tests/mortar/test_meshes/square-ring/conforming_adjacencies:
```diff
@@ -0,0 +1,14 @@
+C  0T  1B
```
I thought that it would be fine since it's just a list of (edge1, edge2) pairs that are conforming, with `T,B,L,R` representing `top, bottom, left, right`, respectively. If the `ispec`s are known, then it's fairly easy to follow.

Adding the position verification step gives another way to recover (at least a good guess of) the true adjacency map, so this test is a little redundant. I have removed it.

int-ptr-ptr committed:
added an assert to adjacency_map.cpp:node_edge()

int-ptr-ptr committed:
found more cases to assert in static helper functions

Rohit-Kakodkar commented:
test this please

int-ptr-ptr committed:
attempt fix: unit-tests/CMakeLists.txt -- add mortar to LINK_DIRS

int-ptr-ptr committed:
Merge branch 'devel' into adjacency2d

int-ptr-ptr committed:
fix: compile error from merge

lsawade commented:
@int-ptr-ptr it would have been great if you had made the recent changes a PR into this branch. That would have made it easier for us to just review the new updates you introduced. The PR is now again at 187 files, and it is becoming quite a behemoth. 

Note, I'm not asking you to do anything, but for next time. Please create significant updates in a separate PR into a main PR becomes it is becoming ungovernable. See, e.g. #1012 and #1014, which are merging into #999. 

There is no extra work because if this branch is merged and delete, the base branch for the extra PR would change to devel.

lsawade commented:
Test this please.

Rohit-Kakodkar commented:
@int-ptr-ptr can you fix nvidia tests.

int-ptr-ptr committed:
fix: test_assembly_mapping to properly use quadrature host space

Rohit-Kakodkar commented:
test this please

int-ptr-ptr committed:
Merge remote-tracking branch 'origin/devel' into adjacency2d

int-ptr-ptr committed:
replaced outdated define_shape_functions call with mesh_impl's

Rohit-Kakodkar commented:
Test this please

int-ptr-ptr commented:
@Rohit-Kakodkar, @lsawade @icui   If it's easier to review, I can reset the PR to when you all did the initial reviews, then git-rebase the remainder into a new PR. We could handle the `devel` incompatibilities in-between.

int-ptr-ptr committed:
revert to commit "changes for PR approval"

hash = be72e9dec2dd7e9ffa27d6aa75b0d41903e8671f

int-ptr-ptr committed:
complete(?) reset to be72e9...

int-ptr-ptr commented:
probably best to do a squash merge onto a temp branch

int-ptr-ptr commented:
> Where are we on this?

I was under the impression that we would merge this PR into a separate branch, then I can handle the remainder in 2 more PRs. Since this branch is equivalent to what you guys have already reviewed, it's just a matter of verifying (through the diff link I provided earlier) that the commits are equivalent.

Once the new branch is made, I can issue PRs into that.

Rohit-Kakodkar commented:
> > Where are we on this?
> 
> I was under the impression that we would merge this PR into a separate branch, then I can handle the remainder in 2 more PRs. Since this branch is equivalent to what you guys have already reviewed, it's just a matter of verifying (through the diff link I provided earlier) that the commits are equivalent.
> 
> Once the new branch is made, I can issue PRs into that.

I just changed the base for this PR. Can you handle the merge conflicts? Then you'd want to create another 2 PR's into `devel-dg` after this is merged?

int-ptr-ptr commented:
> I just changed the base for this PR. Can you handle the merge conflicts? Then you'd want to create another 2 PR's into `devel-dg` after this is merged?
> 
> `devel-dg` is currently on par with `devel`

How would you like me to handle the merge conflicts? Won't that create a bunch of file changes and put us back to square one?

Rohit-Kakodkar commented:
> > I just changed the base for this PR. Can you handle the merge conflicts? Then you'd want to create another 2 PR's into `devel-dg` after this is merged?
> > `devel-dg` is currently on par with `devel`
> 
> How would you like me to handle the merge conflicts? Won't that create a bunch of file changes and put us back to square one?

I am thinking we'd merge this into `devel-dg` after you fix merge conflicts. Then you can issue 2 PRs against the updated `devel-dg`
1. With changes from review comments 
2. With traces with tests 

You'd have to merge the updated `devel-dg` into those 2 PR branches. After those branches have been merged into `devel-dg` we can then create a PR into devel with everything. 

Will that work? Am I missing anything, will this method have too many files to review?

int-ptr-ptr committed:
checkout core, include/enumerations from devel-dg, resolve CMakeLists.txt

int-ptr-ptr committed:
Merge branch 'devel-dg' into adjacency2d

int-ptr-ptr commented:
551 files changed from the conflict resolution -- This is why I thought that this would go on a fresh branch, and the resolution would happen in another PR

Rohit-Kakodkar commented:
I don't understand where the fresh branch is coming from. That fresh branch needs to spawn off a current branch on PU SPECFEMPP repository. 

Am I missing something?

int-ptr-ptr commented:
Stale. PR deprecated by #1044 and #1046.
